File: Kalorie/Kalorie/Features/Moderation/ModerationCatalogueEditorViewModel.swift
//
//  ModerationCatalogueEditorViewModel.swift
//  Kalorie
//
//  Created by Josef Antoni on 10.09.2026.
//

import Foundation
import UIKit

final class ModerationCatalogueEditorViewModel: ObservableObject, NutritionLabelPrefilling {

    // MARK: - Properties

    @Published var barcodeQuery = ""
    @Published private(set) var loadedItem: FoodItemDomain?
    @Published var formInput = FoodItemFormInput()
    @Published private(set) var state: LoadingState<Void> = .idle
    @Published var alertItem: AlertItem?
    @Published private(set) var didSave = false
    @Published var recognizedFields: Set<FoodItemFormField> = []
    @Published var isRecognizingNutritionLabel = false
    @Published var isNutritionLabelCameraVisible = false
    @Published var nutritionLabelCameraHint: String?

    private let fetchFoodItemByBarcode: any FetchFoodItemByBarcodeUseCaseProtocol
    private let updateFoodItem: any UpdateFoodItemUseCaseProtocol
    private let recognizeNutritionLabelUseCase: any RecognizeNutritionLabelUseCaseProtocol
    private let cameraAuthorizationProvider: any CameraAuthorizationProviderProtocol
    private let initialBarcode: String?

    // MARK: - Init

    init(
        fetchFoodItemByBarcode: any FetchFoodItemByBarcodeUseCaseProtocol,
        updateFoodItem: any UpdateFoodItemUseCaseProtocol,
        recognizeNutritionLabel: any RecognizeNutritionLabelUseCaseProtocol,
        cameraAuthorizationProvider: any CameraAuthorizationProviderProtocol,
        initialBarcode: String? = nil
    ) {
        self.fetchFoodItemByBarcode = fetchFoodItemByBarcode
        self.updateFoodItem = updateFoodItem
        self.recognizeNutritionLabelUseCase = recognizeNutritionLabel
        self.cameraAuthorizationProvider = cameraAuthorizationProvider
        self.initialBarcode = initialBarcode
        self.barcodeQuery = initialBarcode ?? ""
    }

    // MARK: - Functions

    @MainActor
    func onAppear() async {
        guard let initialBarcode, loadedItem == nil else { return }
        barcodeQuery = initialBarcode
        await onSearchTapped()
    }

    @MainActor
    func onNutritionLabelCaptured(_ image: UIImage, liveBarcode: String?) async {
        await recognizeNutritionLabel(from: image, liveBarcode: liveBarcode, using: recognizeNutritionLabelUseCase)
    }

    @MainActor
    func onNutritionLabelCameraTapped() async {
        await openNutritionLabelCamera(using: cameraAuthorizationProvider) {
            alertItem = AlertItem(title: L10n.AddFood.cameraPermissionAlert)
        }
    }

    @MainActor
    func onSearchTapped() async {
        guard !barcodeQuery.isEmpty else { return }
        state = .loading
        defer { state = .loaded }
        do {
            guard let item = try await fetchFoodItemByBarcode(barcode: barcodeQuery) else {
                loadedItem = nil
                alertItem = AlertItem(title: L10n.AddFood.errorBarcodeNotFound)
                return
            }
            loadedItem = item
            formInput = FoodItemFormInput(item: item)
            recognizedFields = []
            didSave = false
        } catch {
            Log.error(error, category: Constants.LogCategory.moderation)
            alertItem = AlertItem(title: L10n.Common.errorUnknown)
        }
    }

    @MainActor
    func onSaveTapped() async {
        guard let loadedItem else { return }
        state = .loading
        defer { state = .loaded }
        let item = formInput.asFoodItemDomain(date: loadedItem.date)
        do {
            try await updateFoodItem(item, previouslyLoaded: loadedItem)
            self.loadedItem = item
            didSave = true
        } catch {
            Log.error(error, category: Constants.LogCategory.moderation)
            switch error as? UpdateFoodItemError {
            case .invalidCode:
                alertItem = AlertItem(title: L10n.AddFood.errorInvalidCode)
            case .invalidName:
                alertItem = AlertItem(title: L10n.AddFood.errorInvalidName)
            case .invalidCalories:
                alertItem = AlertItem(title: L10n.AddFood.errorInvalidCalories)
            case .invalidWeight:
                alertItem = AlertItem(title: L10n.AddFood.errorInvalidWeight)
            case .invalidPortion(let portionError):
                alertItem = AlertItem(title: portionError.alertTitle)
            case .changedSinceLoad:
                alertItem = AlertItem(title: L10n.Moderation.errorItemChangedSinceLoad)
            case nil:
                alertItem = AlertItem(title: L10n.Common.errorUnknown)
            }
        }
    }

}

File: Android/app/src/main/java/antoni/kalorie/features/moderation/ModerationCatalogueEditorViewModel.kt
package antoni.kalorie.features.moderation

import androidx.lifecycle.ViewModel
import antoni.kalorie.R
import antoni.kalorie.core.models.FoodItemDomain
import antoni.kalorie.core.usecases.FetchFoodItemByBarcodeUseCaseProtocol
import antoni.kalorie.core.usecases.UpdateFoodItemError
import antoni.kalorie.core.usecases.UpdateFoodItemUseCaseProtocol
import antoni.kalorie.core.utils.AlertItem
import antoni.kalorie.core.utils.Constants
import antoni.kalorie.core.utils.LoadingState
import antoni.kalorie.core.utils.Log
import antoni.kalorie.features.addfoodsheet.FoodItemFormInput
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

class ModerationCatalogueEditorViewModel(
    private val fetchFoodItemByBarcode: FetchFoodItemByBarcodeUseCaseProtocol,
    private val updateFoodItem: UpdateFoodItemUseCaseProtocol,
    private val initialBarcode: String? = null,
) : ViewModel() {

    // MARK: - Properties

    val barcodeQuery = MutableStateFlow(initialBarcode ?: "")
    private val _loadedItem = MutableStateFlow<FoodItemDomain?>(null)
    val loadedItem: StateFlow<FoodItemDomain?> = _loadedItem
    val formInput = MutableStateFlow(FoodItemFormInput())
    private val _state = MutableStateFlow<LoadingState<Unit>>(LoadingState.Idle)
    val state: StateFlow<LoadingState<Unit>> = _state
    val alertItem = MutableStateFlow<AlertItem?>(null)
    private val _didSave = MutableStateFlow(false)
    val didSave: StateFlow<Boolean> = _didSave

    // MARK: - Functions

    suspend fun onAppear() {
        val barcode = initialBarcode ?: return
        if (_loadedItem.value != null) return
        barcodeQuery.value = barcode
        onSearchTapped()
    }

    suspend fun onSearchTapped() {
        val query = barcodeQuery.value
        if (query.isEmpty()) return
        _state.value = LoadingState.Loading
        try {
            val item = fetchFoodItemByBarcode(query)
            if (item == null) {
                _loadedItem.value = null
                alertItem.value = AlertItem(titleRes = R.string.addFood_error_barcodeNotFound)
                return
            }
            _loadedItem.value = item
            formInput.value = FoodItemFormInput.from(item)
            _didSave.value = false
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            Log.error(error, Constants.LogCategory.MODERATION)
            alertItem.value = AlertItem(titleRes = R.string.common_error_unknown)
        } finally {
            _state.value = LoadingState.loaded
        }
    }

    suspend fun onSaveTapped() {
        val loadedItem = _loadedItem.value ?: return
        _state.value = LoadingState.Loading
        val item = formInput.value.asFoodItemDomain(date = loadedItem.date)
        try {
            updateFoodItem(item, loadedItem)
            _loadedItem.value = item
            _didSave.value = true
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            Log.error(error, Constants.LogCategory.MODERATION)
            alertItem.value = AlertItem(
                titleRes = when (error) {
                    is UpdateFoodItemError.InvalidCode -> R.string.addFood_error_invalidCode
                    is UpdateFoodItemError.InvalidName -> R.string.addFood_error_invalidName
                    is UpdateFoodItemError.InvalidCalories -> R.string.addFood_error_invalidCalories
                    is UpdateFoodItemError.InvalidWeight -> R.string.addFood_error_invalidWeight
                    is UpdateFoodItemError.InvalidPortion -> error.error.alertTitleRes
                    is UpdateFoodItemError.ChangedSinceLoad -> R.string.moderation_error_itemChangedSinceLoad
                    else -> R.string.common_error_unknown
                },
            )
        } finally {
            _state.value = LoadingState.loaded
        }
    }
}
